sw/device/silicon_creator/lib/drivers/flash_ctrl.c

```diff
@@ -471,6 +471,155 @@ void flash_ctrl_data_default_cfg_set(flash_ctrl_cfg_t cfg) {
+      sec_mmio_read32(kBase + FLASH_CTRL_DEFAULT_REGION_REG_OFFSET);
+  return (flash_ctrl_cfg_t){
+      .scrambling = bitfield_field32_read(
+          default_region, FLASH_CTRL_DEFAULT_REGION_SCRAMBLE_EN_FIELD),
+      .ecc = bitfield_field32_read(default_region,
+                                   FLASH_CTRL_DEFAULT_REGION_ECC_EN_FIELD),
+      .he = bitfield_field32_read(default_region,
+                                  FLASH_CTRL_DEFAULT_REGION_HE_EN_FIELD),
+  };
+}
+
+// This X macro helps to generate code that operates on each of the flash_ctrl
+// memory protection regions.
+#define FLASH_CTRL_MP_REGIONS(X) \
+  X(0)                           \
+  X(1)                           \
+  X(2)                           \
+  X(3)                           \
+  X(4)                           \
+  X(5)                           \
+  X(6)                           \
+  X(7)
+
+// Defines the bounds of the given memory protection region by setting the
+// MP_REGION_${region} register.
+static void flash_ctrl_mp_region_write(flash_ctrl_region_index_t region,
+                                       uint32_t page_offset,
+                                       uint32_t num_pages) {
+#define FLASH_CTRL_MP_REGION_WRITE_(region)                               \
+  case (region): {                                                        \
+    uint32_t mp_region = FLASH_CTRL_MP_REGION_##region##_REG_RESVAL;      \
+    /* Write the region's base address into the bitfield. */              \
+    mp_region = bitfield_field32_write(                                   \
+        mp_region, FLASH_CTRL_MP_REGION_##region##_BASE_##region##_FIELD, \
+        page_offset);                                                     \
+    /* Write the region's size in pages into the bitfield. */             \
+    mp_region = bitfield_field32_write(                                   \
+        mp_region, FLASH_CTRL_MP_REGION_##region##_SIZE_##region##_FIELD, \
+        num_pages);                                                       \
+    /* Write the bitfield to the MP_REGION_${region} register. */         \
+    sec_mmio_write32(kBase + FLASH_CTRL_MP_REGION_##region##_REG_OFFSET,  \
+                     mp_region);                                          \
+    return;                                                               \
+  }
+
+  switch (region) {
+    FLASH_CTRL_MP_REGIONS(FLASH_CTRL_MP_REGION_WRITE_)
+    default:
+      OT_UNREACHABLE();
+  }
+
+#undef FLASH_CTRL_MP_REGION_WRITE_
+}
+
+// Resets the given region's memory protection config by resetting the
+// MP_REGION_CFG_${region} register, which implicitly disables the region.
+static void flash_ctrl_mp_region_cfg_reset(flash_ctrl_region_index_t region) {
+#define FLASH_CTRL_MP_REGION_CFG_RESET_(region)                               \
+  case (region): {                                                            \
+    static_assert((FLASH_CTRL_MP_REGION_CFG_##region##_REG_RESVAL &           \
```

Comment:
macro arg should not use the same name as the function arg.

right after case:
```
HARDENED_CHECK_EQ(region, MACRO_ARG_REGION);
```